Refactor FeatureBitset and remove helper functions

seelabs opened the pull request:
This PR builds on @scottschurr work refactoring the tests. Only the commits starting with `Refactor FeatureBitset` need to be reviewed.

The motivation for this PR is the proliferation of FeatureBitset helper functions that I think are better served by using bitset primitives. For example, ` hasFeature` should be replaced with `operator[]`, `makeFeatureBitset` and `with_only_features` should be constructors, `supported_features_plus(..)` can be `supported_amendments().set(...)` and `supported_features_except(...)` can be `supported_amendments().reset(...)` (for the one parameter case - the multi parameter case is handled a different way).

mellery451 commented on src/test/jtx/Env_test.cpp:
```diff
@@ -702,7 +703,7 @@ class Env_test : public beast::unit_test::suite
             Env env{
                 *this,
-                with_only_features(
+                FeatureBitset(
```
in-keeping with the other changes, you could just remove `neverSupported` and use `*neverSupportedFeat` directly in a call to `.set()`

seelabs replied:
`set` is not a `const` function, so that would change the collection. However, I did refactor this a bit.

seelabs commented on src/test/app/SetAuth_test.cpp:
```diff
@@ -58,10 +58,7 @@ struct SetAuth_test : public beast::unit_test::suite
-                with_only_features (featureTrustSetAuth)};
-
-            Env env(*this, features | trustSetAuth);
+            Env env(*this, addFeatures(features, featureTrustSetAuth));
```
Fixed (by removing the `addFeatures` functions)

seelabs commented:
Squashed, but not marked as passed yet. This will need to be rebased onto @scottschurr's PR when that PR is finalized.

seelabs commented:
This branch has been rebased onto @scottschurr `minimize use of jtx::with_only_features` branch. That required many more changes than I hopped. I'm hoping that branch won't require too many more changes because rebasing on that is somewhat painful (serves me right for creating a PR on top of another active PR).

@mellery451 @scottschurr I also added two new commits. The first adds some new operators to the `FeatureBitset`. In particular, a "set difference" implemented with `operator-` and I added overloads for operators that take `(Bitset,uint256)` overloads. The second refactors the existing test to take advantage of these new operators. Please review.

scottschurr commented on src/test/app/Offer_test.cpp:
```diff
@@ -3340,7 +3340,7 @@ class Offer_test : public beast::unit_test::suite
         // The problem was identified when featureOwnerPaysFee was enabled,
         // so make sure that gets included.
-        Env env {*this, features | FeatureBitset(featureOwnerPaysFee)};
+        Env env {*this, features | featureOwnerPaysFee};
```
Here is an example of the kind of usage that makes me a little squirmy that we don't supply `+` with the right semantics.  Yes, mathematically, we get the right result.  But there's a cognitive dissonance that I can't use `+` to accomplish the opposite of `-`.

Not a deal breaker.  And we certainly shouldn't change this right now.  And maybe I'll get used to this...

seelabs replied:
We have a couple or operations on our set:
union: `operator|`
intersection: `operator&`
compliment: `operator~`
set difference: `operator-` (`a - b` is (`intersection(a, compliment(b))`)

`operator|` seems like a better choice for union than `operator+`, these are _set_ operations, not arithmetic operations. I wish there were a better choice than `operator-` for set difference, but I can't think of one (`operator\` would work if available).

The best argument I see for allowing `operator+` is in expressions that need to both set and unset features: `all - feat1 + feat2 - feat3` would do the right thing `all - feat1 | feat2 - feat2` not only looks weird, but has the wrong operator precedence. However, we don't actually do that in the code (the vast majority is always either adding or removing features, but not both).

I'd like to live with `operator|` for a while, but yeah, we can revisit adding `operator+` later.

scottschurr commented on src/test/app/Offer_test.cpp:
```diff
@@ -4570,26 +4570,26 @@ class Offer_test : public beast::unit_test::suite
-        FeatureBitset const flow{featureFlow};
-        FeatureBitset const f1373{fix1373};
-        FeatureBitset const flowCross{featureFlowCross};
+        auto const& flow{featureFlow};
```
See comment on Flow_test.cpp.

seelabs replied:
Fixed

seelabs commented on src/ripple/protocol/Feature.h:
```diff
@@ -234,13 +250,70 @@ class FeatureBitset
+        return FeatureBitset{lhs} ^ rhs;
+    }
+
+    // set differenct
```
Fixed

seelabs commented on src/test/app/Flow_test.cpp:
```diff
@@ -1298,28 +1295,28 @@ struct Flow_manual_test : public Flow_test
+        auto const& flow{featureFlow};
+        auto const& f1373{fix1373};
+        auto const& flowCross{featureFlowCross};
+        auto const& f1513{fix1513};
```
Agreed, and fixed (Offer_test.cpp as well).

scottschurr commented on src/test/app/Offer_test.cpp:
```diff
@@ -4519,74 +4517,117 @@ class Offer_test : public beast::unit_test::suite
+        auto const& flow{featureFlow};
+        auto const& f1373{fix1373};
+        auto const& flowCross{featureFlowCross};
+        auto const& f1513{fix1513};
```
These would benefit from being declared as `FeatureBitset const` as well.  And, while you're in there, the many spaces following the declaration of local `all` on lines 4579 and 4601 no longer align with anything.  We might as well lose those spaces.

seelabs replied:
Fixed

scottschurr commented:
:+1:

seelabs commented:
Rebased onto @scottschurr's `subtract-features` branch and squashed

seelabs commented:
rebased onto 0.80.1

seelabs commented:
In 0.90.0-b1
